controller/system_backup_controller.go

```diff
@@ -0,0 +1,1068 @@
+		defer func() {
+			if err != nil {
+				c.cacheStateUpdate(
+					systemBackupRecordTypeError, longhorn.SystemBackupStateError,
+					longhorn.SystemBackupConditionReasonUpload, SystemBackupErrSync,
+				)
+				return
+			}
+
+			c.cacheStateUpdate(
+				systemBackupRecordTypeNormal, longhorn.SystemBackupStateReady,
+				EventReasonSynced, SystemBackupMsgSyncedBackupTarget,
+			)
+		}()
+
+		systemBackupCfg, err := backupTargetClient.SystemBackupConfigGet(systemBackup.Name, longhornVersion)
+		if err != nil {
+			return err
+		}
+
+		systemBackup.Status.Version = systemBackupCfg.LonghornVersion
+		systemBackup.Status.GitCommit = systemBackupCfg.LonghornGitCommit
+		systemBackup.Status.ManagerImage = systemBackupCfg.ManagerImage
+		systemBackup.Status.CreatedAt = metav1.Time{Time: systemBackupCfg.CreatedAt}
+		systemBackup.Status.LastSyncedAt = metav1.Time{Time: time.Now().UTC()}
+
+		return nil
+	}
+
+	archievePath := filepath.Join(SystemBackupTempDir, systemBackup.Name+types.SystemBackupExtension)
+	tempDir := filepath.Join(SystemBackupTempDir, systemBackup.Name)
+
+	switch systemBackup.Status.State {
+	case longhorn.SystemBackupStateNone:
+		if err = c.InitSystemBackup(systemBackup, log); err != nil {
+			return err
+		}
+
+		c.cacheStateUpdate(
+			systemBackupRecordTypeNormal, longhorn.SystemBackupStateGenerate,
+			EventReasonStart, SystemBackupMsgStarting,
+		)
+
+	case longhorn.SystemBackupStateGenerate:
+		go c.GenerateSystemBackup(systemBackup, archievePath, tempDir)
+
+	case longhorn.SystemBackupStateUpload:
+		go c.UploadSystemBackup(systemBackup, archievePath, tempDir, backupTargetClient)
+
+	case longhorn.SystemBackupStateReady:
+		cleanupLocalSystemBackupFiles(archievePath, tempDir, log)
+
+	case longhorn.SystemBackupStateError:
+		cleanupLocalSystemBackupFiles(archievePath, tempDir, log)
+
+	case longhorn.SystemBackupStateDelete:
+		err = c.cleanupRemoteSystemBackupFiles(systemBackup, backupTargetClient)
+		if err != nil {
+			c.cacheStateUpdate(
+				systemBackupRecordTypeError, longhorn.SystemBackupStateError,
+				longhorn.SystemBackupConditionReasonDelete, SystemBackupErrDelete,
+			)
+			return err
+		}
+
+		cleanupLocalSystemBackupFiles(archievePath, tempDir, log)
+
+		err = c.ds.RemoveFinalizerForSystemBackup(systemBackup)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *SystemBackupController) InitSystemBackup(systemBackup *longhorn.SystemBackup, log logrus.FieldLogger) error {
+	systemBackup.Status.Version = meta.Version
+	systemBackup.Status.GitCommit = meta.GitCommit
+	return nil
+}
+
+func (c *SystemBackupController) UploadSystemBackup(systemBackup *longhorn.SystemBackup, archievePath, tempDir string, backupTargetClient engineapi.BackupTargetClientSystemBackupInterface) {
+	log := getLoggerForSystemBackup(c.logger, systemBackup.Name, backupTargetClient.GetURL())
+
+	var recordErr error
+	existingSystemBackup := systemBackup.DeepCopy()
+	defer func() {
```

Comment:
NIT: Add a comment mentioning that this function is executed by a separate goroutine hence the CR update should be handled here.

Reply:
Updated.